### PR TITLE
ld64-latest: build 450.3

### PR DESCRIPTION
**Pull request by kencu**
* add man pages if ld64-latest is installed
* add build fixes for older systems
* clarify compiler blacklisting for some older ld64 versions

closes: https://trac.macports.org/ticket/57587
closes: https://trac.macports.org/ticket/53784

###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

most systems macOS 10.4 to 10.14
Xcode 2.5 to 11

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?

- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by kencu** [comment]
I would appreciate anyone interested in installing this and testing it.
I patched out the i386 deprecation warning to prevent errors on older systems.
The Availabilty Macros in dyld.h are tricky -- I mostly just defined them all to nothing. Any insight or help there would be appreciated.

I am having some trouble trying to fully grok how to run the test suite without Xcode, and I would appreciate any clever persons who understand how to do this to help out with my education in that regard.

This builds ld64-latest @450.3 on 10.6.8 and newer, and with the testing to date, seems functional and without errors. I tweaked the newer dyld.h to fix the build of ld64-97 on 10.4 PPC. More testing with more systems still needed.

**Comment by kencu** [comment]
Makefile-351 and Makefile-409 can be deleted once we decide we're going to skip right over those two versions (likely). I just left them in for now as remnants of the iterative process that got us to here.

**Comment by kencu** [comment]
@ryandesign @MarcusCalhoun-Lopez @jmroot -- I would appreciate it if you would try to kick this half to death and see if it holds up for you.

**Comment by kencu** [comment]
hmmm. There is still some work to do on that API_UNAVAILABLE stuff, perhaps just in dyld_private.h. 

Also, looks like there is a circular dependency happening on Xcode 7.3 <https://paste.macports.org/180025acd6aa>  where to build ld64-latest something is calling in ld64-latest... I am not sure just now how I'm going to get past that.

**Comment by kencu** [comment]
Pretty good with this plan; just need to see about Xcode 7.3 (10.11). Looks like it is circular on ld64-latest due to calling in clang-9.0 which calls in ld64-latest. Possibly some earlier systems will show this as well. Can either fix the build with the Xcode clang, if possible, or alternatively change the default ld64 on those systems to ld64-274.

All the systems 10.6 and up can build and use ld64-latest, once they have built a bootstrap linker.

**Comment by kencu** [comment]
looks like the issue is with libtapi, which needs a clang version > 801 to build. So on < 10.12 macports therefore calls in macports-clang-9.0 to do the build, which has a runtime dep on ld64, and I think it is this that sets up the circular dependency issue.

File: devel/ld64/Portfile [hunk 1]
```diff
@@ -171,6 +181,51 @@ subport ld64-latest {
...
+        ld64-ppc-9610466.patch \
+        ld64-409-snapshot-no-mkpath-np.diff \
+        ld64-409-lto-file-llvm-3.4-fix.diff \
+        ld64-409-Options-strndup.diff \
+        ld64-409-add-missing-machine-defs.diff \
+        ld64-409-options-disable-i386-warning.diff \
+        ld64-450-move-baseplatform-def-to-header.diff \
+
+# below have been reassessed against ld64-450.3 only
+# NOT NEEDED       ld64-134-missing-include.patch \
```
Comment (jeremyhu) [inline devel/ld64/Portfile]: I'd leave these out.

File: devel/ld64/Portfile [hunk 2]
```diff
@@ -187,18 +242,24 @@ if {${subport} eq ${name}} {
     variant ld64_97 conflicts ld64_127 ld64_236 ld64_xcode description {Use ld64-97 as the default linker (last version that works on Tiger)} {}
```
Comment (jeremyhu) [inline devel/ld64/Portfile]: Update conflicts.

File: devel/ld64/files/Makefile-351 [hunk 3]
```diff
@@ -0,0 +1,106 @@
+ifdef LLVM_CONFIG
```
Comment (jeremyhu) [inline devel/ld64/files/Makefile-351]: Don't need this one, it seems.

File: devel/ld64/files/Makefile-409 [hunk 4]
```diff
@@ -0,0 +1,107 @@
+ifdef LLVM_CONFIG
```
Comment (jeremyhu) [inline devel/ld64/files/Makefile-409]: Don't need this one, it seems.

**Comment by kencu** [comment]
Thanks. Running the test suite on ld64-latest:

To perform the test suite, I manually did something like this:
```
Running ld64 unit tests:

cd into 

ld64-450.3/unit-tests/test-cases

run:

perl ../bin/make-recursive.pl CXX="/opt/local/bin/clang++-mp-9.0 -arch i386" CC="/opt/local/bin/clang-mp-9.0 -arch i386" BUILT_PRODUCTS_DIR="/opt/macports-ports/devel/ld64/work/ld64-450.3/" LD_SYSROOT="/" LD="/opt/macports-ports/devel/ld64/work/ld64-450.3/ld" OSX_SDK="/" ARCH="i386"  | perl ../bin/result-filter.pl
perl ../bin/make-recursive.pl CXX="/opt/local/bin/clang++-mp-9.0 -arch x86_64"  CC="/opt/local/bin/clang-mp-9.0 -arch x86_64" BUILT_PRODUCTS_DIR="/opt/macports-ports/devel/ld64/work/ld64-450.3/" LD_SYSROOT="/" LD="/opt/macports-ports/devel/ld64/work/ld64-450.3/ld" OSX_SDK="/" ARCH="x86_64" | perl ../bin/result-filter.pl

or raw unfiltered results:

perl ../bin/make-recursive.pl CXX="/opt/local/bin/clang++-mp-9.0 -arch i386" CC="/opt/local/bin/clang-mp-9.0 -arch i386" BUILT_PRODUCTS_DIR="/opt/macports-ports/devel/ld64/work/ld64-450.3/" LD_SYSROOT="/" LD="/opt/macports-ports/devel/ld64/work/ld64-450.3/ld" OSX_SDK="/" ARCH="i386"
perl ../bin/make-recursive.pl CXX="/opt/local/bin/clang++-mp-9.0 -arch x86_64"  CC="/opt/local/bin/clang-mp-9.0 -arch x86_64" BUILT_PRODUCTS_DIR="/opt/macports-ports/devel/ld64/work/ld64-450.3/" LD_SYSROOT="/" LD="/opt/macports-ports/devel/ld64/work/ld64-450.3/ld" OSX_SDK="/" ARCH="x86_64"


to clean tests and run again:

perl ../bin/make-recursive.pl clean
```

The results on 10.7.5 were fairly good, I thought:
```
$ perl ../bin/make-recursive.pl CXX="/opt/local/bin/clang++-mp-9.0 -arch x86_64"  CC="/opt/local/bin/clang-mp-9.0 -arch x86_64" BUILT_PRODUCTS_DIR="/opt/macports-ports/devel/ld64/work/ld64-450.3/" LD_SYSROOT="/" LD="/opt/macports-ports/devel/ld64/work/ld64-450.3/ld" OSX_SDK="/" ARCH="x86_64" | perl ../bin/result-filter.pl

alt-entry                                FAIL Makefile failure
blank-stubs                              FAIL Makefile failure
data-in-code                             FAIL Makefile failure
dead_strip-initializers                  FAIL Makefile failure
demangle                                 FAIL Makefile failure
dwarf-debug-notes-r                      FAIL Makefile failure
dwarf-debug-notes-lto                    FAIL Makefile failure
dwarf-strip                              FAIL spurious stderr failure: warning: no debug symbols in executable (-arch x86_64)
dylib-main                               FAIL Makefile failure
empty-object                             FAIL Makefile failure
init-order                               FAIL Makefile failure
linker-optimization-hints                FAIL Makefile failure
llvm-integration                         FAIL Makefile failure
lto-llvm-options                         FAIL Makefile failure
lto-symbol-section-move                  FAIL Makefile failure
lto-weak-native-override                 FAIL Makefile failure
objc-category-class-property-mismatch    FAIL Makefile failure
special-labels                           FAIL Makefile failure
stack_size_no_addr                       FAIL Makefile failure
stripped-indirect-symbol-table           FAIL spurious stderr failure: /opt/local/libexec/llvm-9.0/bin/llvm-objdump: warning: section '__DATA,__nl_symbol_ptr' mentioned in a -j/--section option, but not found in any input file
weak_import-disable                      FAIL Makefile failure
 * * * 281 of 302 unit-tests passed (93.0 percent) * * *
```

I will see if I can run the same test suite on 10.14, and sort out if these few failures are important or not...

**Comment by kencu** [comment]
10.14 also shows a few failures, a bit different than 10.14:
```
Executing:  cd "/opt/local/var/macports/build/_opt_macports-ports_devel_ld64/ld64-latest/work/ld64-450.3/unit-tests/test-cases" && perl ../bin/make-recursive.pl CXX="/usr/bin/clang++ -arch x86_64" CC="/usr/bin/clang -arch x86_64" BUILT_PRODUCTS_DIR="/opt/local/var/macports/build/_opt_macports-ports_devel_ld64/ld64-latest/work/ld64-450.3" LD_SYSROOT="/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk" LD="/opt/local/var/macports/build/_opt_macports-ports_devel_ld64/ld64-latest/work/ld64-450.3/ld" OSX_SDK="/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk" ARCH="x86_64" | perl ../bin/result-filter.pl 
data-in-code                             FAIL Makefile failure
blank-stubs                              FAIL Makefile failure
dwarf-debug-notes-lto                    FAIL Makefile failure
weak_import-local                        FAIL Makefile failure
dwarf-debug-notes-r                      FAIL Makefile failure
re-export-optimizations                  FAIL Makefile failure
dwarf-strip                              FAIL spurious stderr failure: warning: no debug symbols in executable (-arch x86_64)
lto-symbol-section-move                  FAIL Makefile failure
re-export-layers                         FAIL Makefile failure
re-export-optimizations-indirect         FAIL Makefile failure
demangle                                 FAIL Makefile failure
re-export-cases                          FAIL Makefile failure
lto-llvm-options                         FAIL Makefile failure
custom-segment-layout                    FAIL Makefile failure
init-order                               FAIL Makefile failure
dylib-re-export-cycle                    FAIL Makefile failure
install-name-override                    FAIL Makefile failure
weak_dylib                               FAIL Makefile failure
symbol-hiding-umbrella                   FAIL Makefile failure
empty-object                             FAIL Makefile failure
dylib-main                               FAIL Makefile failure
linker-optimization-hints                FAIL Makefile failure
stripped-indirect-symbol-table           FAIL spurious stderr failure: /opt/local/libexec/llvm-9.0/bin/llvm-objdump: warning: section '__DATA,__nl_symbol_ptr' mentioned in a -j/--section option, but not found in any input file
symbol-moving                            FAIL Makefile failure
llvm-integration                         FAIL Makefile failure
lto-weak-native-override                 FAIL Makefile failure
objc-category-class-property-mismatch    FAIL Makefile failure
 * * * 272 of 299 unit-tests passed (91.0 percent) * * *
```

nice to get 100%... but this looks reasonable I think, barring any genius like @Ionic explaining to me how to fix these test errors...

**Comment by kencu** [comment]
OK. Will commit today.

**Pull request merged**